### ISSUE TITLE
Contradictory advice and test on using Test::More

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13231]
Migrated from [rt.perl.org#119619](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=119619) (status was 'open')

Searchable as RT119619$

**Comment by p5pRT** [comment]
### From @Smylers


This is a bug report for perl from smylers@&#8203;stripey\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.19\.4\.




---
perlhack\, in its TESTING section\, says that tests other than in specific
subdirectories should use Test&#8203;::More \(rather than t/test\.pl\)\.

But t/porting/test\_bootstrap\.t specifically tests that no t/\*\*\.t files
use Test&#8203;::More — adding the line&#8203;:

&nbsp;  use Test&#8203;::More;

into a test in t/porting/ \(not one of the directories perlhack mentions
as requiring ad\-hoc testing or test\.pl\) makes test\_bootstrap\.t fail\.

These obviously contradict each other\, and I don't know what should be
used for a new test in t/porting/\.

Either test\_bootstrap\.t needs restricting to the directories mentioned
in perlhack\, or the advice in perlhack needs changing to say never to
use Test&#8203;::More\.



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.19\.4&#8203;:

Configured by smylers at Tue Sep  3 10&#8203;:58&#8203;:31 BST 2013\.

Summary of my perl5 \(revision 5 version 19 subversion 4\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 5565c73819a0eed045991803240cb7122b86b800
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.8\.0\-30\-generic\, archname=x86\_64\-linux
&nbsp;    uname='linux fozzie 3\.8\.0\-30\-generic \#44\-ubuntu smp thu aug 22 20&#8203;:52&#8203;:24 utc 2013 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-des \-Dusedevel'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.7\.3'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.17'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector'




---
@&#8203;INC for perl 5\.19\.4&#8203;:
&nbsp;    lib
&nbsp;    /home/smylers/lib/perl5/site\_perl
&nbsp;    /home/smylers/lib/perl5
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.19\.4/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.19\.4
&nbsp;    /usr/local/lib/perl5/5\.19\.4/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/5\.19\.4
&nbsp;    \.



---
Environment for perl 5\.19\.4&#8203;:
&nbsp;    HOME=/home/smylers
&nbsp;    LANG=en\_GB\.utf8
&nbsp;    LANGUAGE=en\_GB&#8203;:en
&nbsp;    LC\_COLLATE=C
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/smylers/bin&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/sbin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/usr/games
&nbsp;    PERL5LIB=/home/smylers/lib/perl5/site\_perl&#8203;:/home/smylers/lib/perl5
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_CPANM\_OPT=\-\-sudo \-\-prompt
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @jkeenan

On Thu Sep 05 04&#8203;:44&#8203;:55 2013\, smylers@&#8203;stripey\.com wrote&#8203;:
> 
> This is a bug report for perl from smylers@&#8203;stripey\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.19\.4\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> perlhack\, in its TESTING section\, says that tests other than in
> specific
> subdirectories should use Test&#8203;::More \(rather than t/test\.pl\)\.
> 
> But t/porting/test\_bootstrap\.t specifically tests that no t/\*\*\.t files
> use Test&#8203;::More — adding the line&#8203;:
> 
>   use Test&#8203;::More;
> 
> into a test in t/porting/ \(not one of the directories perlhack
> mentions
> as requiring ad\-hoc testing or test\.pl\) makes test\_bootstrap\.t fail\.
> 
> These obviously contradict each other\, and I don't know what should be
> used for a new test in t/porting/\.
> 
> Either test\_bootstrap\.t needs restricting to the directories mentioned
> in perlhack\, or the advice in perlhack needs changing to say never to
> use Test&#8203;::More\.
> 

I can confirm that adding or editing a file in t/porting which contains
'use Test&#8203;::More' will cause t/porting/test\_bootstrap\.t to fail\.

Would the patch attached suffice?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Add-t-porting-to-list-of-directories-which-use-t-tes.patch</summary>

```diff
From c0f09e3f907fe356fdf7d7ac36c9bea024894e0d Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Fri, 6 Sep 2013 01:12:30 +0200
Subject: [PATCH] Add t/porting to list of directories which use t/test.pl.

For: RT #119619
---
 pod/perlhack.pod |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/pod/perlhack.pod b/pod/perlhack.pod
index ff685d2..028849c 100644
--- a/pod/perlhack.pod
+++ b/pod/perlhack.pod
@@ -754,7 +754,7 @@ tested.  Tests in F<t/opbasic>, for instance, have been placed there
 rather than in F<t/op> because they test functionality which
 F<t/test.pl> presumes has already been demonstrated to work.
 
-=item * F<t/cmd>, F<t/run>, F<t/io> and F<t/op>
+=item * F<t/cmd>, F<t/run>, F<t/io>, F<t/op> and F<t/porting>
 
 Now that basic require() and subroutines are tested, you can use the
 F<t/test.pl> library.
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Smylers

James E Keenan via RT writes&#8203;:

> On Thu Sep 05 04&#8203;:44&#8203;:55 2013\, smylers@&#8203;stripey\.com wrote&#8203;:
> 
> > perlhack\, in its TESTING section\, says that tests other than in
> > specific subdirectories should use Test&#8203;::More \(rather than
> > t/test\.pl\)\.
> > 
> > But t/porting/test\_bootstrap\.t specifically tests that no t/\*\*\.t
> > files use Test&#8203;::More — adding the line&#8203;:
> > 
> >   use Test&#8203;::More;
> > 
> > into a test in t/porting/ \(not one of the directories perlhack
> > mentions as requiring ad\-hoc testing or test\.pl\) makes
> > test\_bootstrap\.t fail\.

Note that t/porting/ is simply an example which demonstrates the
contradiction; the problem isn't restricted to t/porting\. My apologies
if that wasn't clear in my bug report\.

Here's the relevant text from perlhack&#8203;:

&nbsp;  ·   t/base\, t/comp and t/opbasic

&nbsp;      \.\.\. use ad hoc tests for these three \.\.\.
&nbsp;      
&nbsp;  ·   t/cmd\, t/run\, t/io and t/op

&nbsp;      Now that basic require\(\) and subroutines are tested\, you can use
&nbsp;      the t/test\.pl library\. \.\.\.

&nbsp;  ·   Everything else

&nbsp;      Now that the core of Perl is tested\, Test&#8203;::More can and should be
&nbsp;      used\. \.\.\.

And here are t/ subdirectories which include \.t files but aren't
mentioned in either of the first two bullet points above&#8203;:

&nbsp;  lib x2p benchmark mro bigmem porting uni japh re test\_pl win32


> > These obviously contradict each other\, and I don't know what should
> > be used for a new test in t/porting/\.
> > 
> > Either test\_bootstrap\.t needs restricting to the directories
> > mentioned in perlhack\, or the advice in perlhack needs changing to
> > say never to use Test&#8203;::More\.
> 
> I can confirm that adding or editing a file in t/porting which
> contains 'use Test&#8203;::More' will cause t/porting/test\_bootstrap\.t to
> fail\.
> 
> Would the patch attached suffice?

That patch simply moves t/porting into the second bullet point's list\,
which obviously doesn't resolve the contradiction in general\.

As far as I can tell\, test\_bootstrap\.t prevents any test in t/ from
using Test&#8203;::More\, and as such the ‘everything else’ bullet point in
perlhack can't ever apply\.

So making the documentation consistent with test\_bootstrap\.t would
require entirely removing that third bullet point which mentions
Test&#8203;::More\, and renaming the second one to say ‘everything else’ \(that
is\, everything but the three files mentioned in the first bullet point\)\.

But it isn't clear to me that that's the right thing to do\. The author
of that advice intended that Test&#8203;::More could be used in some
circumstances\, and the reasoning sound plausible to me\. As such\, it may
be that adjusting test\_bootstrap\.t to match the advice \(only prohibiting
Test&#8203;::More selectively\) would be better\.

So before any solution can be effected\, somebody needs to make a
judgement call about what's desired here\.

Cheers

Smylers
\-\- 
Stop drug companies hiding negative research results\.
Sign the AllTrials petition to get all clinical research results published\.
Read more&#8203;: http&#8203;://www\.alltrials\.net/blog/the\-alltrials\-campaign/

**Comment by Smylers** [comment]
This contradiction still exists in blead: [perlhack](/Perl/perl5/blob/blead/pod/perlhack.pod#testing) says to use `Test::More`, but if you do, [test_bootstrap.t](/Perl/perl5/blob/blead/t/porting/test_bootstrap.t#L40) will object that you have done so.

Given tests seem to be existing without using `Test::More`, shall I submit a doc patch reversing the advice and saying not to use it?

**Comment by jkeenan** [comment]
> This contradiction still exists in blead: [perlhack](/Perl/perl5/blob/blead/pod/perlhack.pod#testing) says to use `Test::More`, but if you do, [test_bootstrap.t](/Perl/perl5/blob/blead/t/porting/test_bootstrap.t#L40) will object that you have done so.
> 
> Given tests seem to be existing without using `Test::More`, shall I submit a doc patch reversing the advice and saying not to use it?

Since the last time we discussed this issue was more than six years ago, I'll have to re-read the ticket to respond adequately.

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
> > This contradiction still exists in blead: [perlhack](/Perl/perl5/blob/blead/pod/perlhack.pod#testing) says to use `Test::More`, but if you do, [test_bootstrap.t](/Perl/perl5/blob/blead/t/porting/test_bootstrap.t#L40) will object that you have done so.
> > Given tests seem to be existing without using `Test::More`, shall I submit a doc patch reversing the advice and saying not to use it?
> 
> Since the last time we discussed this issue was more than six years ago, I'll have to re-read the ticket to respond adequately.
> 
> Thank you very much.
> Jim Keenan

I've now re-read this ticket and believe I understand your points.  In the first patch in the `jkeenan/ghi-13231-test-more-advice` branch and in a p.r. to follow, I have modified `pod/perlhack.pod` to reflect this.

In the course of demonstrating to myself the correctness of your argument, I found one gotcha:  a test under `t/lib` which, though it did not `use Test::More`, did presume that `Test::` modules could be loaded via `use MODULE`.  The test file in question said `use Test::Simple`.  So in the second patch in the branch I modify that file `require t/test.pl` and adapt `t/porting/test_bootstrap.t` accordingly.

Please review.

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
See: https://github.com/Perl/perl5/pull/17524

**Comment by jkeenan** [comment]
Closed vi commits 915b794bfdc480ffedc696439aa063ae4191d5e3 and 
070720ff0f9b7ca6e49d9f9d2780d90b05aa31c4
